src/gpu/generic/sycl/rnn/rnn_utils.hpp

```diff
@@ -0,0 +1,447 @@
+/*******************************************************************************
+* Copyright 2019-2024 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef GPU_INTEL_OCL_RNN_RNN_UTILS_HPP
+#define GPU_INTEL_OCL_RNN_RNN_UTILS_HPP
+
+#include "common/c_types_map.hpp"
+#include "common/memory_desc_wrapper.hpp"
+#include "common/memory_storage.hpp"
+#include "common/memory_tracking.hpp"
+#include "common/primitive_desc.hpp"
+#include "common/stream.hpp"
+#include "gpu/generic/sycl/sycl_gpu_primitive.hpp"
+
+inline int calc_4d_off(int i0, int i1, int d1, int i2, int d2, int i3, int d3,
+        int i4, int d4) {
+    return ((((i0) * (d1) + (i1)) * (d2) + (i2)) * (d3) + (i3)) * (d4) + (i4);
+}
+
+#define cell_execution_sig(f) \
+    status_t f(impl::engine_t *engine, const exec_ctx_t &ctx, dim_t dir, \
+            dim_t lay, dim_t iter, const rnn_utils::user_data_t &user_data, \
+            const rnn_utils::workspace_t &workspace, \
+            const rnn_utils::scratch_t &scratch) const
+
+#define grid_execution_sig(f) \
+    status_t f(impl::engine_t *engine, const exec_ctx_t &ctx, \
+            const rnn_utils::user_data_t &user_data, \
+            const rnn_utils::workspace_t &workspace, \
+            const rnn_utils::scratch_t &scratch) const
+
+#define gemm_sig(f) \
+    status_t f(impl::engine_t *engine, const exec_ctx_t &ctx, \
+            std::unique_ptr<memory_storage_t> &a, \
+            std::unique_ptr<memory_storage_t> &b, \
+            std::unique_ptr<memory_storage_t> &c, gemm_kind_t gemm_kind) const
```

Comment:
Please consider removing these defines, they do not generally align with the style used within oneDNN. Rather than duplicating complicated interfaces with `#define`, we should be able to get effectively the same simplification by passing a function context structure which supports aggregate initialization similar to the following example.

```C++
// Context Structure
struct func_ctx_t {
    int arg1;
    int arg2;
}

//Function interface
status_t func(const func_ctx_t &ctx)

//Sample call
func({a, b});
```